scarb/src/compiler/plugin/proc_macro/compilation.rs

```diff
@@ -0,0 +1,35 @@
+use crate::compiler::plugin::proc_macro::PROC_MACRO_BUILD_PROFILE;
+use crate::core::{Config, Package};
+use crate::flock::Filesystem;
+use libloading::library_filename;
+
+/// This trait is used to define the shared library path for a package.
+pub trait SharedLibraryProvider {
+    /// Location of Cargo `target` directory.
+    fn target_path(&self, config: &Config) -> Filesystem;
+    /// Location of the shared library for the package.
+    fn shared_lib_path(&self, config: &Config) -> Filesystem;
+}
+
+impl SharedLibraryProvider for Package {
+    fn target_path(&self, config: &Config) -> Filesystem {
+        let ident = format!("{}-{}", self.id.name, self.id.source_id.ident());
+        // Defines the Cargo target directory in cache, as:
+        // `/(..)/SCARB_CACHE/plugins/proc_macro/<package_name>-<source_id_ident>/target/`
+        config
+            .dirs()
+            .procedural_macros_dir()
+            .into_child(ident)
+            .into_child("target")
+    }
```

Comment:
If you have two projects that you work on simultaneously, and one is using plugin `X v1` and another `X v2`, then this logic (to my understanding) will cause Scarb to always recompile the plugin when you switch context.

I'd consider including package version here. I know that this potentially will greately increase cache size, but we can always implement some kind of cache gc, like Cargo recently did.